internal/config/etcd_source.go

```diff
@@ -0,0 +1,161 @@
+// Licensed to the LF AI & Data foundation under one
+// or more contributor license agreements. See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/milvus-io/milvus/internal/log"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.uber.org/zap"
+)
+
+const (
+	ModeWatch = iota
+	ModeInterval
+)
+
+type EtcdSource struct {
+	etcdCli *clientv3.Client
+	ctx     context.Context
+	sync.RWMutex
+	currentConfig map[string]string
+
+	keyPrefix       string
+	refreshMode     int
+	refreshInterval time.Duration
+	eh              EventHandler
+}
+
+func NewEtcdSource(remoteInfo *EtcdInfo) (*EtcdSource, error) {
+	etcdCli, err := clientv3.New(clientv3.Config{
+		Endpoints:   remoteInfo.Endpoints,
+		DialTimeout: 5 * time.Second,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &EtcdSource{
+		etcdCli:         etcdCli,
+		ctx:             context.Background(),
+		keyPrefix:       remoteInfo.KeyPrefix,
+		refreshMode:     remoteInfo.RefreshMode,
+		refreshInterval: remoteInfo.RefreshInterval,
+	}, nil
+}
+
+// Delete implements ConfigSource
+func (es *EtcdSource) Delete(key string) error {
+	return nil
+}
+
+// GetConfigurationByKey implements ConfigSource
+func (es *EtcdSource) GetConfigurationByKey(key string) (string, error) {
+	es.RLock()
+	v, ok := es.currentConfig[key]
+	es.RUnlock()
+	if !ok {
+		return "", fmt.Errorf("key not found: %s", key)
+	}
+	return v, nil
+}
+
+// GetConfigurations implements ConfigSource
+func (es *EtcdSource) GetConfigurations() (map[string]string, error) {
+	configMap := make(map[string]string)
+	err := es.refreshConfigurations()
+	if err != nil {
+		return nil, err
+	}
+	if es.refreshMode == ModeInterval {
+		go es.refreshConfigurationsPeriodically()
+	}
```

Comment:
Each call of `GetConfigurations()` will start a goroutine to refreshConfigurations. Maybe we need a "sync.Once" protection?